### PR TITLE
8. Time in a single field (Hours and minutes)

### LINKED ISSUE #7: Supp time calculation
When calculating the supp time, the result is sometimes wrong. When the result should be 4, it shows 56...

### PR DESCRIPTION
**Pull request by allan-brunner**
Have hours and minutes in a single field instead of 2 separated.
Dynamic required time also made in this pull request

fix [issue 7](https://github.com/ZeaCold/TimeCalculator/issues/7)

### CODE REVIEW

File: MainWindow.xaml.cs [hunk 1]
```diff
@@ -1,58 +1,251 @@
...
+        /// and write specfic character
+        /// </summary>
+        /// <param name="sender">The textbox that triggered the event</param>
+        /// <param name="e">Args of the key event</param>
+        private void Field_PreviewKeyDown(object sender, KeyEventArgs e)
+        {
+            TextBox textBox = sender as TextBox;
+
+            // marque l'�v�nement comme g�r�
+            e.Handled = true;
```
Comment (clementfavre) [inline MainWindow.xaml.cs]: Commentary in French among all the others in English

File: MainWindow.xaml.cs [hunk 2]
```diff
@@ -1,58 +1,251 @@
 using System.Collections.Generic;
+using System.Text.RegularExpressions;
 using System.Windows;
 using System.Windows.Controls;
 using System.Windows.Input;
 
+using ZC.TimeCalculator.Resources.Extension;
```
Comment (clementfavre) [inline MainWindow.xaml.cs]: Don't forget to run a code cleanup before each commit to avoid this kind of excess space

File: MainWindow.xaml.cs [hunk 3]
```diff
@@ -1,58 +1,251 @@
...
 {
     /// <summary>
     /// Logique d'interaction pour MainWindow.xaml
     /// </summary>
     public partial class MainWindow : Window
     {
-        const int M_In_H = 60;                  // Constant value of minutes in an hour
-        const int H_In_D = 24;                  // Constant value of minutes in a day
+        const int M_In_H = 60;                              // Constant value of minutes in an hour
+        const string DefaultText = "__:__";                 // Constant value of the default text
```
Comment (clementfavre) [inline MainWindow.xaml.cs]: How about yesterday, shouldn't the constants be in full case?

**Review by clementfavre — CHANGES_REQUESTED** [review]
There are changes to be made.

And there's an additional case to deal with in terms of times: when the user finishes after midnight, it's noted, for example, at 26:50, which is impossible :
image

File: MainWindow.xaml.cs (hunk 2 above)
Comment (allan-brunner) [reply MainWindow.xaml.cs]: It was on purpose to separate usings of System and ZeaCold but I can unite them if it's better

File: MainWindow.xaml [hunk 4]
```diff
@@ -1,30 +1,184 @@
...
-        <TextBox x:Name="txtMEndSupp" HorizontalAlignment="Left" Margin="350,72,0,0" TextWrapping="Wrap" VerticalAlignment="Top" Width="50" Height="26" HorizontalContentAlignment="Center" VerticalContentAlignment="Center" IsReadOnly="True"/>
+        <Label
+            x:Name="lblRequired"
+            Width="60"
+            Margin="9,10,0,0"
+            HorizontalAlignment="Left"
+            VerticalAlignment="Top"
+            HorizontalContentAlignment="Right"
+            VerticalContentAlignment="Center"
+            Content="Required" />
```
Comment (clementfavre) [inline MainWindow.xaml]: Text in English among others in French

File: Resources/Utils/TimeUtils.cs [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+using System.Text.RegularExpressions;
+using System.Windows.Controls;
+using System.Windows.Input;
+using ZC.TimeCalculator.Resources.Extension;
+
+namespace ZC.TimeCalculator.Resources.Utils
+{
+    public class TimeUtils
+    {
+        public const int MINUTES_IN_HOUR = 60;                                                          // Constant value of minutes in an hour
```
Comment (clementfavre) [inline Resources/Utils/TimeUtils.cs:12]: Unused ://

File: Resources/Utils/TimeUtils.cs [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+        public const int HOURS_IN_DAY = 24;                                                             // Constant value of hours in a day
+
+        public static Regex timeFormat = new Regex("(-?)([_01][_0-9]|[_2][_0-3]):([_0-5][_0-9])");      // Regex for the default time format
```
Comment (clementfavre) [inline Resources/Utils/TimeUtils.cs:16]: Unused in this MainWindows.xaml.cs but re-created and used in the Utils file.

**Review by clementfavre — CHANGES_REQUESTED** [review]
There's still the same bug reported in the old review, except in one other case

image

File: Resources/Utils/TimeUtils.cs (hunk 5 above)
Comment (allan-brunner) [reply Resources/Utils/TimeUtils.cs:12]: MINUTES_IN_HOUR in TimeUtils is used. Wasn't used in MainWindow.xaml.cs

Comment (clementfavre) [reply Resources/Utils/TimeUtils.cs:12]: yes, I took the wrong example

**Review by clementfavre — APPROVED** [review]
Tested and approved! I haven't found any hour bugs (yet (●'◡'●)), GJ !

**Pull request merged**